src/languageservice/httpClient.ts

```diff
@@ -0,0 +1,143 @@
+
+   /*
+    * Downloads a file and stores the result in the temp file inside the package object
+    */
+    public downloadFile(urlString: string, pkg: IPackage, logger: ILogger, statusView: IStatusView, proxy?: string, strictSSL?: boolean): Promise<void> {
+        const url = parseUrl(urlString);
+        let options = this.getHttpClientOptions(url, proxy, strictSSL);
+        let client = url.protocol === 'http:' ? http : https;
+
+        return new Promise<void>((resolve, reject) => {
+            if (!pkg.tmpFile || pkg.tmpFile.fd === 0) {
+                return reject(new PackageError('Temporary package file unavailable', pkg));
+            }
+
+            let request = client.request(options, response => {
+                if (response.statusCode === 301 || response.statusCode === 302) {
+                    // Redirect - download from new location
+                    return resolve(this.downloadFile(response.headers.location, pkg, logger, statusView, proxy, strictSSL));
+                }
+
+                if (response.statusCode !== 200) {
+                    // Download failed - print error message
+                    logger.appendLine(`failed (error code '${response.statusCode}')`);
+                    return reject(new PackageError(response.statusCode.toString(), pkg));
+                }
+
+                // If status code is 200
+                this.handleSuccessfulResponse(pkg, response, logger, statusView).then(_ => {
+                    resolve();
+                }).catch(err => {
+                    reject(err);
+                });
+            });
+
+            request.on('error', error => {
+                reject(new PackageError(`Request error: ${error.code || 'NONE'}`, pkg, error));
+            });
+
+            // Execute the request
+            request.end();
+        });
+    }
+
+    private getHttpClientOptions(url: Url, proxy?: string, strictSSL?: boolean): any {
+        const agent = getProxyAgent(url, proxy, strictSSL);
+
+        let options: http.RequestOptions = {
+            host: url.hostname,
+            path: url.path,
+            agent: agent,
+            port: +url.port
+        };
+
+        if (url.protocol === 'https:') {
+            let httpsOptions: https.RequestOptions = {
+                    host: url.hostname,
+                    path: url.path,
+                    agent: agent,
+                    port: +url.port
+            };
+            options = httpsOptions;
+        }
+
+        return options;
+    }
+
+   /*
+    * Calculate the download percentage and stores in the progress object
+    */
+    public handleDataReceivedEvent(progress: IDownloadProgress, data: any, logger: ILogger, statusView: IStatusView): void {
+        progress.downloadedBytes += data.length;
+
+        // Update status bar item with percentage
+        let newPercentage = Math.ceil(100 * (progress.downloadedBytes / progress.packageSize));
```

Comment:
I can add a check  for 0 just in case

---
In reply to: [93141773](https://github.com/Microsoft/vscode-mssql/pull/557#discussion_r93141773) [](ancestors = 93141773)